src/devices/CharacterLcd/LcdInterface.Spi.cs

```diff
@@ -0,0 +1,175 @@
+                    }
+                    else
+                    {
+                        return 0;
+                    }
+                }
+            }
+
+            private void Initialize()
+            {
+                // This sequence (copied from a python example) completely resets the display (if it was
+                // previously erroneously used with 8 bit access, it may not return to normal operation otherwise)
+                SendCommandAndWait(0x3);
+                SendCommandAndWait(0x3);
+                SendCommandAndWait(0x3);
+                SendCommandAndWait(0x2);
+            }
+
+            public override void SendCommand(byte command)
+            {
+                Write4Bits((byte)(0x0 | MapCommandToDataPins((byte)(command >> 4))));
+                Write4Bits((byte)(0x0 | MapCommandToDataPins(command)));
+            }
+
+            public override void SendCommands(ReadOnlySpan<byte> commands)
+            {
+                foreach (byte command in commands)
+                {
+                    SendCommand(command);
+                }
+            }
+
+            public override void SendData(byte value)
+            {
+                Write4Bits((byte)(_registerSelectPin | MapCommandToDataPins((byte)(value >> 4))));
+                Write4Bits((byte)(_registerSelectPin | MapCommandToDataPins(value)));
+            }
+
+            public override void SendData(ReadOnlySpan<byte> values)
+            {
+                foreach (byte value in values)
+                {
+                    SendData(value);
+                }
+            }
+
+            public override void SendData(ReadOnlySpan<char> values)
+            {
+                foreach (byte value in values)
+                {
+                    SendData(value);
+                }
+            }
+
+            private void Write4Bits(byte command)
+            {
+                _shiftRegister.ShiftByte((byte)(command | _enablePin | BacklightFlag));
+                _shiftRegister.ShiftByte((byte)((command & ~_enablePin) | BacklightFlag));
+            }
+
+            // This method iterates through the bits of the 'original' command and sets them to the
+            // appropriate position so that they are shifted out to the correct pins.
+            private byte MapCommandToDataPins(byte command)
+            {
+                byte bits = 0x0;
+
+                for (var i = 0; i < _dataPins.Length; i++)
+                {
+                    var bit = (command >> i) & 1;
+
+                    if (bit == 1)
+                    {
+                        bits = (byte)(bits | _dataPins[i]);
+                    }
+                }
+
+                return bits;
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                if (_shouldDispose)
+                {
+                    _shiftRegister?.Dispose();
+                    _shiftRegister = null!;
```

Comment:
This should be outside of the _shoudDispose test. Setting the _shiftRegister to null even without Dispose, will all the GC to clean the variable even if the original class is not.
I know this is coming from the I2C4bits implementation. There are few like this which should be corrected. Btw, you can correct them as well if possible.

Reply:
Should be fixed now.